WebdriverIO project generator for Serenity/JS

jan-molak opened the pull request:
## Proposed changes

Per my discussion with @christian-bromann, I'd like to extend the WebdriverIO project wizard to generate Serenity/JS projects, too.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

christian-bromann reviewed (approved):
This PR seems to only contains improvements for unit tests and some wording. No signs of Serenity in here. If this is intentional, let's go ahead and merge this as it a good patch itself.

LGTM 👍

jan-molak commented:
@christian-bromann - I'm adding Serenity now :)

jan-molak commented:
@christian-bromann - I added project generators for Serenity/JS with Cucumber, Jasmine, and Mocha. Please let me know if you're happy with the overall structure and integration points.

I still need to think about additional tests I could write to ensure that EJS template rendering works as expected.

christian-bromann commented on packages/wdio-cli/src/templates/wdio.conf.tpl.ejs:
```diff
@@ -232,7 +251,13 @@ if (answers.isUsingTypeScript) {
     //
     <%- include('snippets/reporters', { reporters: answers.reporters }) %>
-    <% if(answers.framework === 'mocha') { %>
+    <% if(answers.serenityAdapter === 'mocha') { %>//
```
Can we move all the framework opts into their own template snippets?

jan-molak replied:
Yeah, good call; let me do this now.

jan-molak replied:
Done in https://github.com/webdriverio/webdriverio/pull/11368/commits/68a437c2e7894a647e9ff03e277fc8ad046dcf0e#diff-cc58b213dd871b45469e751b91589bd4836c606b4dbb8af89fc9cc38782fec66

christian-bromann reviewed (commented):
One comment, otherwise LGTM

What do you think @erwinheitzman @tamil777selvan ?

erwinheitzman reviewed (approved):
LGTM 👍

jan-molak commented on packages/wdio-cli/src/commands/config.ts:
```diff
@@ -66,7 +67,8 @@ export const parseAnswers = async function (yes: boolean): Promise<ParsedAnswers
         ...pluginPackages.map(plugin => plugin.package),
-        ...servicePackages.map(service => service.package)
+        ...servicePackages.map(service => service.package),
+        ...serenityPackages,
```
I think so because `servicePackages` are determined based on what the developer selects in the configuration wizard, `serenityPackages` are separate from that concept. Is that what you meant?

erwinheitzman replied:
I see it now, I was on my phone and it was a bit hard to read and I read serenity twice but it's service and serenity.

christian-bromann commented on packages/wdio-cli/src/templates/wdio.conf.tpl.ejs:
```diff
@@ -231,58 +235,14 @@ if (answers.isUsingTypeScript) {
-        ignoreUndefinedDefinitions: false
-    },
-    <% } %>
+    <%- include('snippets/reporters', { reporters: answers.reporters })
+    %><%- answers.serenityAdapter   === 'mocha'    ? include('snippets/serenity-js/mocha', { answers }) : ''
+    %><%- answers.framework         === 'mocha'    ? include('snippets/mocha', { answers }) : ''
+    %><%- answers.serenityAdapter   === 'jasmine'  ? include('snippets/serenity-js/jasmine', { answers }) : ''
+    %><%- answers.framework         === 'jasmine'  ? include('snippets/jasmine', { answers }) : ''
+    %><%- answers.serenityAdapter   === 'cucumber' ? include('snippets/serenity-js/cucumber', { answers }) : ''
+    %><%- answers.framework         === 'cucumber' ? include('snippets/cucumber', { answers }) : ''
+    %>
```
Why splitting between Serenity and standard framework opts? Are they differ by a lot?

jan-molak replied:
They differ quite a bit for Cucumber since S/JS supports all the major versions of Cucumber, which means supporting legacy parameters, tag filters from before tag expressions, and so on.

For Mocha and Jasmine, most configuration options are the same, but there are some slight differences, like support for `strict` flag in S/JS Mocha, but not in WDIO, or support for `cleanStack` in WDIO but not in S/JS
- https://serenity-js.org/api/jasmine-adapter/interface/JasmineConfig/ vs https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-jasmine-framework/src/types.ts
- https://serenity-js.org/api/mocha-adapter/interface/MochaConfig/ vs https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-mocha-framework/src/types.ts

Thinking about it, we could try and standardise the adapters in the future 🤔

jan-molak replied:
Sure, I can do that.

tamil777selvan reviewed (commented):
I have a minor suggestion. Could we have a single template instead of both `js.ejs` and `ts.ejs` as like we currently have?

jan-molak commented:
> I have a minor suggestion. Could we have a single template instead of both `js.ejs` and `ts.ejs` as like we currently have?

That's a good idea; I'll need to introduce some helpers to avoid having to repeat `if` statements.

jan-molak commented:
@christian-bromann and @tamil777selvan - thanks for your feedback. I merged WebdriverIO and Serenity/JS configuration templates. I've also removed separate templates for CJS/ESM and instead generate TS, CJS and ESM from TypeScript templates.

Serenity/JS code sample templates are reasonably involved as they contain classes, functions, interfaces, parameters, return types, and so on. To avoid polluting WDIO EJS templates with a bazillion `if` statements, I introduced an `EjsHelpers` class that helps to generate different kinds of files from the same TS file.

The best example of using the helpers can be seen at https://github.com/webdriverio/webdriverio/pull/11368/files#diff-d2b3a22b9c707586c3011b43acc439b12e3c952a0bf5f70f58cf257558406908

You'll also find unit tests demonstrating the behaviour of `EjsHelpers` at https://github.com/webdriverio/webdriverio/pull/11368/files#diff-99bb565c721ebef00c7ccea7cab374547e3c4a25a78d03a0e321c888aaefec5c

I think I've arrived at quite an elegant solution that could be used to simplify other WDIO templates too if needed.

Looking forward to hearing your thoughts!

christian-bromann reviewed (approved):
LGTM 👍

Great work!

tamil777selvan reviewed (approved):
LGTM 👍
